rts/motoko-rts/src/types.rs

```diff
@@ -256,63 +257,120 @@ impl Value {
     }
 
     /// Assumes that the value is a pointer and returns the pointer value. In debug mode panics if
-    /// the value is not a pointer.
+    /// the value is not a pointer or if it points to a stale object that has been forwarded.
     pub fn get_ptr(self) -> usize {
         debug_assert!(self.get().is_ptr());
         unskew(self.0 as usize)
     }
 
-    /// Get the object tag. In debug mode panics if the value is not a pointer.
+    /// Check that the forwarding pointer is valid.
+    #[inline]
+    pub unsafe fn check_forwarding_pointer(self) {
+        debug_assert!(
+            self.forward().get_ptr() == self.get_ptr()
+                || self.forward().forward().get_ptr() == self.forward().get_ptr()
+        );
+    }
+
+    /// Check whether the object's forwarding pointer refers to a different location.
+    pub unsafe fn is_forwarded(self) -> bool {
+        self.check_forwarding_pointer();
+        self.forward().get_ptr() != self.get_ptr()
```

Comment:
I also thought about this. This is very good question. My motivation was to simplify the changes in the compiler, as the forwarding pointer has the same encoding as the non-forwarded pointers. There are several places in the compiler where it passes (skewed) pointers (especially also in the quite complex serialization logic). So, I indeed traded simplicity for performance. But this could be something for the backlog to optimize.